File: Python/0206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

#define new Linkedlist浪費內存空間
#先定義cur指針，指向head、再定義pre指針，initialize為null
#把cur -> next節點用temp指針保存，將其反轉為指向pre並cycle
#Two-Pointer Method

class Solution(object):
    def reverseList(self, head):
        cur = head
        pre = None
        while cur:
            temp = cur.next #cur的下一個節點要反轉為pre
            cur.next = pre

            #updata pre & cur pointer
            pre = cur
            cur = temp
        return pre

File: Java/0206. Reverse Linked List.java
//We can use 2 pts and recursion
//2 pts

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// 若再 define 一個 新的 Linked List 是對內存空間的浪費
// 只要改變 next 指針的指向，反轉即可
// 先定義一個 cur 指針指向 head，再定義一個 pre 指針初始化為 null，把 cur -> next 用 temp 保存

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode cur = head;
        ListNode temp = null;
        while (cur != null){
            temp = cur.next; //保存下一個節點
            cur.next = prev;
            prev = cur;
            cur = temp;
        }
        return prev;
    }
}